lib/df-detection.c

```diff
@@ -0,0 +1,209 @@
+		current->num_read = 0;
+		current->ranges_size = 0;
+		kfree(current->dfetch_addresses);
+		current->dfetch_addresses = NULL;
+	}
+}
+
+void report(void)
+{
+	int i;
+	unsigned long *first_entries;
+	unsigned int first_nr_entries;
+	unsigned long *second_entries;
+	unsigned int second_nr_entries;
+	unsigned long first_frame = 0;
+	unsigned long second_frame = 0;
+
+	for (i = 0; i < current->dfetch_index; i++) {
+		if (current->dfetch_addresses[current->dfetch_pairs[i].first].stack &&
+		    current->dfetch_addresses[current->dfetch_pairs[i].second].stack) {
+			first_nr_entries = stack_depot_fetch(
+			    current->dfetch_addresses[current->dfetch_pairs[i].first].stack,
+			    &first_entries);
+			second_nr_entries = stack_depot_fetch(
+			    current->dfetch_addresses[current->dfetch_pairs[i].second].stack,
+			    &second_entries);
+			pr_err("BUG: multi-read\n");
+			pr_err("==================================================================\n");
+			dump_stack_print_info(KERN_DEFAULT);
+			pr_err("First Stack Trace:");
+			stack_trace_print(first_entries, first_nr_entries, 0);
+			pr_err("Second Stack Trace:");
+			stack_trace_print(second_entries, second_nr_entries, 0);
+		} else {
+			pr_err("BUG: multi-read\n");
+			pr_err("==================================================================\n");
+		}
+		pr_err(
+		    "First %px len %lu Caller %pSR \nSecond %px len "
+		    "%lu Caller %pSR \n \n",
+		    current->dfetch_addresses[current->dfetch_pairs[i].first].start_address,
+		    current->dfetch_addresses[current->dfetch_pairs[i].first].len,
+		    current->dfetch_addresses[current->dfetch_pairs[i].first].caller,
+		    current->dfetch_addresses[current->dfetch_pairs[i].second].start_address,
+		    current->dfetch_addresses[current->dfetch_pairs[i].second].len,
+		    current->dfetch_addresses[current->dfetch_pairs[i].second].caller);
+		pr_err("==================================================================\n");
+		if (panic_on_warn) {
+			panic_on_warn = 0;
+			panic("panic_on_warn set. \n");
+		}
+	}
+}
+
+depot_stack_handle_t dfetch_save_stack(gfp_t flags)
+{
+	unsigned long entries[STACK_DEPTH];
+	unsigned int nr_entries;
+
+	nr_entries = stack_trace_save(entries, ARRAY_SIZE(entries), 0);
+	nr_entries = filter_irq_stacks(entries, nr_entries);
+	return stack_depot_save(entries, nr_entries, flags);
+}
+
+int is_intersect(struct dfetch_address_range a, struct dfetch_address_range b,
+		 void *kernel_addr)
+{
+	void *a_end = (void *)((char *)a.start_address + a.len);
+	void *b_end = (void *)((char *)b.start_address + b.len);
+	if (a.start_address <= b.start_address && a_end > b.start_address) {
+		size_t len = (char *)(a_end > b_end ? b_end : a_end) -
+			     (char *)b.start_address;
+		get_random_bytes(kernel_addr, len);
+		return 1;
+	} else if (b.start_address <= a.start_address &&
+		   b_end > a.start_address) {
+		unsigned long diff =
+		    (char *)a.start_address - (char *)b.start_address;
```

Comment:
Please insert a blank line after `diff` declaration.
Moving this declaration to the top is optional.